lib/matplotlib/tests/test_figure.py

```diff
@@ -566,3 +566,207 @@ def test_add_subplot_twotuple():
     assert ax4.get_subplotspec().colspan == range(0, 2)
     with pytest.raises(IndexError):
         fig.add_subplot(3, 2, (6, 3))
+
+
+class TestCollage:
+    @check_figures_equal(extensions=['png'])
+    @pytest.mark.parametrize('x', [
+        [["A", "A", "B"], ["C", "D", "B"]],
+        [[1, 1, 2], [3, 4, 2]]]
+    )
+    def test_basic(self, fig_test, fig_ref, x):
+        grid_axes = fig_test.quilt_layout(x)
+
+        for k, ax in grid_axes.items():
+            ax.set_title(k)
+
+        labels = sorted(np.unique(x))
+
+        assert len(labels) == len(grid_axes)
+
+        gs = fig_ref.add_gridspec(2, 3)
+        axA = fig_ref.add_subplot(gs[:1, :2])
+        axA.set_title(labels[0])
+
+        axB = fig_ref.add_subplot(gs[:, 2])
+        axB.set_title(labels[1])
+
+        axC = fig_ref.add_subplot(gs[1, 0])
+        axC.set_title(labels[2])
+
+        axD = fig_ref.add_subplot(gs[1, 1])
+        axD.set_title(labels[3])
+
+    @check_figures_equal(extensions=['png'])
+    def test_all_nested(self, fig_test, fig_ref):
+        x = [["A", "B"], ["C", "D"]]
+        y = [["E", "F"], ["G", "H"]]
+
+        fig_ref.set_constrained_layout(True)
+        fig_test.set_constrained_layout(True)
+
+        grid_axes = fig_test.quilt_layout([[x, y]])
+        for ax in grid_axes.values():
+            ax.set_title(ax.get_label())
+
+        gs = fig_ref.add_gridspec(1, 2)
+        gs_left = gs[0, 0].subgridspec(2, 2)
+        for j, r in enumerate(x):
+            for k, label in enumerate(r):
+                fig_ref.add_subplot(gs_left[j, k]).set_title(label)
+
+        gs_right = gs[0, 1].subgridspec(2, 2)
+        for j, r in enumerate(y):
+            for k, label in enumerate(r):
+                fig_ref.add_subplot(gs_right[j, k]).set_title(label)
+
+    @check_figures_equal(extensions=['png'])
+    def test_nested(self, fig_test, fig_ref):
+
+        fig_ref.set_constrained_layout(True)
+        fig_test.set_constrained_layout(True)
+
+        x = [["A", "B"], ["C", "D"]]
+
+        y = [["F"], [x]]
```

Comment:
OK, so what happens if I do `y = [["F" [x]]["G", [x]]]`?  where the x subgrid spans the two rows?  Not saying you have to support that, but it is valid to do `gs_n = GridSpecFromSubGridspec(gs[:, 1], nrows=2, ncols=2)`.

Reply:
It fails with

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-d094223040d8> in <module>
----> 1 fig.subplot_grid(y)

~/source/p/matplotlib/lib/matplotlib/figure.py in Figure.subplot_grid(self, layout, subplot_kw, gridspec_kw, empty_sentinel)
   1703 rows, cols = layout.shape
   1704 gs = self.add_gridspec(rows, cols, **gridspec_kw)
-> 1705 ret = _do_layout(gs, layout, *_identify_keys_and_nested(layout))
   1706 for k, ax in ret.items():
   1707     if isinstance(k, str):

~/source/p/matplotlib/lib/matplotlib/figure.py in Figure.subplot_grid.<locals>._do_layout(gs, layout, unique_ids, nested)
   1695     if set(output) & set(nested_output):
   1696         overlap = set(output) & set(nested_output)
-> 1697         raise ValueError(f"There are duplicate keys {overlap} "
   1698                          f"between the outer layout\n{layout!r}\n"
   1699                          f"and the nested layout\n{nested_layout}")
   1700     output.update(nested_output)
   1701 return output

ValueError: There are duplicate keys {'A', 'B', 'D', 'C'} between the outer layout
array([['F', list([[['A', 'B'], ['C', 'D']]])],
       ['G', list([[['A', 'B'], ['C', 'D']]])]], dtype=object)
and the nested layout
[[list(['A', 'B']) list(['C', 'D'])]]
```

because the two nested layouts have overlapping keys.